### ISSUE TITLE
Cannot import blacksheep.contents. Can't deploy on Vercel

### ISSUE THREAD

**Issue by clementpoiret** [Neoteroi/BlackSheep#275]
**Describe the bug**
I'm trying to replace my FastAPI apps on Vercel with BlackSheep.

Here is an example: https://github.com/clementpoiret/blacksheep-vercel

Unfortunately, it seems I can't deploy it, it leads to a server error because:

```sh
> Ready! Available at http://0.0.0.0:3000
> Building @vercel/python:server/api.py
Installing required dependencies...
> Built @vercel/python:server/api.py [20s]
Traceback (most recent call last):
  File "/home/cp264607/.cache/com.vercel.fun/runtimes/python3/../python/bootstrap.py", line 147, in <module>
    lambda_runtime_main()
  File "/home/cp264607/.cache/com.vercel.fun/runtimes/python3/../python/bootstrap.py", line 127, in lambda_runtime_main
    fn = lambda_runtime_get_handler()
  File "/home/cp264607/.cache/com.vercel.fun/runtimes/python3/../python/bootstrap.py", line 113, in lambda_runtime_get_handler
    mod = importlib.import_module(module_name)
  File "/home/cp264607/anaconda3/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/tmp/zeit-fun-b57c65a91f69d/vc__handler__python.py", line 12, in <module>
    __vc_spec.loader.exec_module(__vc_module)
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "./server/api.py", line 1, in <module>
    from blacksheep import Application, Content, Response
  File "/tmp/zeit-fun-b57c65a91f69d/blacksheep/__init__.py", line 7, in <module>
    from .contents import Content as Content
ModuleNotFoundError: No module named 'blacksheep.contents'
TypeError: Object prototype may only be an Object or null: undefined
    at Function.setPrototypeOf (<anonymous>)
    at new LambdaError (/usr/lib/node_modules/vercel/dist/index.js:9679:16)
    at Lambda.<anonymous> (/usr/lib/node_modules/vercel/dist/index.js:9812:27)
    at Generator.next (<anonymous>)
    at fulfilled (/usr/lib/node_modules/vercel/dist/index.js:9726:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Do you know what is happening?
Thanks

**Comment by RobertoPrevato** [comment]
Hi @clementpoiret 
I apologize for replying late - I was very busy this week. I am not familiar with Vercel, therefore I did a bit of reading.
Judging by the comments I read under this blog post: https://dev.to/andrewbaisden/how-to-deploy-a-python-flask-app-to-vercel-2o5k

I suspect that it's possible to deploy a Python app with all dependencies (deploying the virtual environment packages without building them on the Vercel server) - this is not going to work with custom extensions libraries like those of BlackSheep. I just signed-up for Vercel and I'll give it a try, to gain a better insight. Thanks for sharing the project template!

**Comment by RobertoPrevato** [comment]
I investigated for more than one hour, I learnt some basics of Vercel and run an application locally, but I couldn't find yet why the BlackSheep import fails.

**Comment by clementpoiret** [comment]
Hi @RobertoPrevato, thanks for your investigation. Yes it also fails locally, so I think we can discard the "Vercel OS hypothesis". Mine was that vercel is looking for .py files only, which aren't here. Maybe its not compatible with cpython ?

**Comment by RobertoPrevato** [comment]
I managed to make some progress!

Here you can see an API I deployed to Vercel, where you get a reply from BlackSheep: https://blacksheep-vercel-six.vercel.app - the import error is caused by exotic things Vercel does with the Python packages.
I found a fix by accidentally pushing a change, but this should fix the import error:

```python
import os
import sys

sys.path.append(os.getcwd())
```

The error page from BlackSheep is caused by another problem: the Vercel server doesn't trigger the ASGI server lifecycle events - currently BlackSheep requires the "Start-up" event being fired by the ASGI server.
I might change this detail because it's causing several problems when underlying servers don't trigger the start event, like in this case.

The problem on the local environment happens for another reason, which I can explain tomorrow. Now it's 11 PM here. 😄

**Comment by clementpoiret** [comment]
Ohh nice finding! Don't you think we can open an issue on vercel also? Because if it happens with blacksheep it can also happen for other libs 🤔

**Comment by RobertoPrevato** [comment]
> Don't you think we can open an issue on vercel also? Because if it happens with blacksheep it can also happen for other libs 🤔

Judging from what I saw yesterday night, I think so. It seems that relative imports are not working because they move Python library files and they (probably) should copy the extension files (`.so` files in this case) to the root folder. This being said, I'll try to create a dev distribution package of BlackSheep that doesn't use relative imports, and instead always uses absolute imports. This might fix the problem.

When it comes to the problem that happens locally, it seems it's caused by the fact that `vercel dev` doesn't use the same Python interpreter that is available when running the `vercel dev` command itself:

- I prepared a Python 3.10 virtual environment
- `vercel dev` uses Python 3.8 on the host
- Python 3.8 doesn't have the extensions available because they depend on the version of Python

(this might be beause vercel doesn't support Python 3.10, yet?) - I'm not sure.

**Comment by RobertoPrevato** [comment]
The other problem regarding lifecycle events might also be caused by breaking changes in `uvicorn` or the `ASGI` specification: I need to check this before I can be certain that it's a problem in the Vercel ASGI handler.

If it doesn't handle life cycle events in general, this is indeed an issue because web applications often need to handle disposable things such as HTTP connection pools, DB connection pools upon start and stop.

**Comment by RobertoPrevato** [comment]
I confirm my findings about the local environment. I made an experiment to confirm the case:

1. `vercel dev` uses different Python interpreters when installing dependencies and then when running the function app: in my case, Python 3.10 is used when installing dependencies, meaning that extensions get installed for this version of Python (e.g. `example.cpython-310-x86_64-linux-gnu.so`)
2. later when trying to run the application, it uses Python 3.8 from `/usr/bin/python3` - this fails because the extension that were installed are for a different version of Python - they should be `example.cpython-38-x86_64-linux-gnu.so`

if I copy the required files for Python 3.8 at runtime before trying to import, it fixes the problem, but indeed the error happens for `httptools` - which is also using C extensions.

In my opinion this should be submitted as a bug to Vercel. Right now I don't want to put more time into this, though, because I have very little time. 😢 

In summary, we are speaking about 3 different problems in this issue.

**Comment by clementpoiret** [comment]
Thanks a lot, you already made awesome progress! I'll open issues directly for Vercel and reference this one :)
